### PR TITLE
#187354251  Update / changing user password

### PR DESCRIPTION
**Pull request by P-Rwirangira**
## Purpose

The purpose of the following PR is to help the user to change his/her password but not when forgotten the first password 

## Changes Made

-Added a controller in userController.ts for updating/changing user password
-Added API endpoint for updating/changing password in userRouter.js
-Added tests

## Testing Instructions

Check on localhost:{PORT}/users/update-password (On Postman use POST method)

## Related Issues

No issues at the moment

## Checklist

Please review the following checklist and make sure all tasks are complete before submitting:

- [x] Code follows the project's coding standards
- [x] Changes are covered by tests
- [x] Documentation is updated (if applicable)
- [x] All tests pass

### CODE REVIEW

File: src/controllers/userController.ts [hunk 1]
```diff
@@ -33,5 +34,38 @@ const createUser = async (req: Request, res: Response): Promise<void> => {
     res.status(500).send('Internal Server Error');
   }
 };
+const updatePassword = async (req: Request, res: Response): Promise<void> => {
+  try {
+    const { oldPassword, newPassword, email } = req.body;
```
Comment (niyontwali) [inline src/controllers/userController.ts]: You cannot request an email from a user who is already signed in. Instead, explore methods for changing or updating the password using various resources. Retrieve the email from the token or from the request object appended by an authentication middleware.

File: src/controllers/userController.ts [hunk 2]
```diff
@@ -33,5 +34,38 @@ const createUser = async (req: Request, res: Response): Promise<void> => {
...
+    const hashedNewPassword = await bcrypt.hash(newPassword, saltRound);
+    await User.update(
+      { password: hashedNewPassword },
+      {
+        where: {
+          email: email,
+        },
+      }
+    );
+    res.status(200).json({ message: 'Successfully updated user password!' });
```
Comment (niyontwali) [inline src/controllers/userController.ts]: Well done, but please ensure you include ok: true or ok: false in the response. For example
```javascript
 res.status(400).json({
    ok: false,
    message: 'User not found'
 );
 ```
 
Please avoid using `send()` and use `json()` instead

File: src/controllers/userController.ts [hunk 3]
```diff
@@ -33,5 +34,38 @@ const createUser = async (req: Request, res: Response): Promise<void> => {
...
+  } catch (error) {
+    logger.error('Error updating user:', error);
+    res.status(500).send('Internal Server Error');
```
Comment (niyontwali) [inline src/controllers/userController.ts]: Please use the reusable function sendInternalErrorResponse from src/validations. Refer to how other controller functions have utilized it.

File: src/docs/users.yml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+              newPassword:
+                type: string
+              email:
+                type: string
+      responses:
+        '200':
+          description: Successfully updated user password
+          schema:
+            type: object
+            properties:
```
Comment (niyontwali) [inline src/docs/users.yml]: This should be done in the auth.yml

File: src/routes/userRoute.ts [hunk 5]
```diff
@@ -1,9 +1,9 @@
...
 import { Router } from 'express';
-import { createUser } from '../controllers/userController';
+import { createUser, updatePassword } from '../controllers/userController';
 
 const router = Router();
 
 router.post('/', createUser);
-
+router.put('/update-password', updatePassword);
 export default router;
```
Comment (niyontwali) [inline src/routes/userRoute.ts]: These should be in th authRoute.ts

**Review by niyontwali — CHANGES_REQUESTED** [review]

File: src/controllers/userController.ts (hunk 1 above)
Comment (P-Rwirangira) [reply src/controllers/userController.ts]: Changed it to extracting id from the token

File: src/controllers/userController.ts (hunk 2 above)
Comment (P-Rwirangira) [reply src/controllers/userController.ts]: Done

File: src/controllers/userController.ts (hunk 3 above)
Comment (P-Rwirangira) [reply src/controllers/userController.ts]: I changed to using the reusable fucntions

File: src/docs/users.yml (hunk 4 above)
Comment (P-Rwirangira) [reply src/docs/users.yml]: Done

File: src/routes/userRoute.ts (hunk 5 above)
Comment (P-Rwirangira) [reply src/routes/userRoute.ts]: Done

**Review by hozayves — CHANGES_REQUESTED** [review]
Yes Patrick 

Great work on the PR! Just a couple of suggestions:

1. It might be beneficial to move the updatePassword controller into the user controller file for better code organization.
2. Consider capturing the email from the logged-in user using the auth middleware from @JeanIrad  instead of from the request body. This can enhance security and align with best practices for handling user data.

**Review by hozayves — APPROVED** [review]

**Pull request closed without merging**